library_management_system_downloader/run_api_download_and_parse_results.py

```diff
@@ -0,0 +1,201 @@
+               String, ForeignKey("dois_table.database_id"), nullable=False),
+        Column('timestamp', DateTime, nullable=False,
+               default=func.now()),  # Default is creation time
+        Column('xml_response', String, nullable=False),
+        Column('full_text_indicator', Integer))
+
+dois_table.create(
+        checkfirst=True)  # Don't create the table if it already exists.
+
+library_holdings_table.create(
+        checkfirst=True)  # Don't create the table if it already exists.
+
+# Create an sqlalchemy portal for inserting new records into
+# the tables:
+dois_table_insert = dois_table.insert()
+holdings_table_insert = library_holdings_table.insert()
+
+# =============================================================================
+# Define functions for querying and inserting into the database
+# =============================================================================
+
+Session = sessionmaker(bind=sql_engine)
+sql_session = Session()
+
+
+def is_doi_already_answered_in_database(
+        doi,
+        sqlalchemy_session=sql_session):
+    existing_doi_with_fulltext_anwer = sql_session.query(distinct(
+        dois_table.c.doi)).join(library_holdings_table).filter(
+            dois_table.c.doi == doi).filter(
+                    library_holdings_table.c.full_text_indicator.isnot(
+                            None)).first()
+
+    if existing_doi_with_fulltext_anwer is None:
+        return False
+    else:
+        return True
+
+# Example invocation:
+# is_doi_already_answered_in_database('10.1001/2012.jama.10139')
+
+
+def insert_a_doi_database_record(
+        doi,
+        api_response_text,
+        full_text_indicator_for_doi,
+        sqlalchemy_session=sql_session):
+    doi_foreign_key = sql_session.query(
+            distinct(dois_table.c.database_id)).filter(
+                    dois_table.c.doi == doi).first()
+
+    if doi_foreign_key is None:
+        logging.info(f'Inserting DOI "{doi}" into the database...')
+        inserted_doi = dois_table_insert.execute(doi=doi)
+        # Get the foreign key we just inserted into the database:
+        doi_foreign_key = inserted_doi.inserted_primary_key[0]
+    else:
+        doi_foreign_key = doi_foreign_key[0]
+        logging.info(f'DOI "{doi}" is already in the database (foreign key '
+                     '"{doi_foreign_key}"), so not inserting it...')
+
+    # Insert the new XML record:
+    logging.info('Inserting holdings record into database...')
+    holdings_table_insert.execute(
+            doi_foreign_key=doi_foreign_key,
+            xml_response=api_response_text,
+            full_text_indicator=full_text_indicator_for_doi)
+
+# =============================================================================
+# Import the list of DOIs
+# =============================================================================
+
+# Import the datset into
+doi_dataset = pd.read_csv(config.tsv_dataset_location, sep='\t', header=0)
```

Comment:
`pd.read_table` defaults to `sep='\t'`. `header` not necessary.

Reply:
I'm fine taking out both. Having said that, since TSV is (in my experience) less-used than CSV, it makes sense to me to keep this explicit, so that anyone using a CSV might more quickly get the idea to change that (whether there or in a comment, I imagine that *I'd* like the reminder in the future). With that in mind, would you be ok with keeping it in?

Reply:
The point is you're reading a TSV so you should use the function that defaults to tab-separation.

For the crossref and scihub repos, we're using TSV as the tabular plain text format. I'd recommend continuing with TSV output (always with column names) for this repo. Many of our fields may contain commas, but will rarely contain tabs.

Reply:
I didn't see in your initial comment that you were suggesting replacing `read_csv` with `read_table`. I've done that in 930fc14, removed the default arguments, and added a comment clarifying that a TSV is expected.